src/ocean/math/WideInt.d

```diff
@@ -0,0 +1,474 @@
+            b.payload[i] = 1337;
+        }
+        test!("==")(a, b);
+        a.payload[$-1] = 0;
+        test!("!=")(a, b);
+    }
+
+    /***************************************************************************
+
+        Increment current value by one fitting in uint
+
+        Params:
+            rhs = value to increment with. Limited to `uint` for now to simplify
+                internal arithmetic.
+
+        Throws:
+            WideIntRangeException if this WideInt value range was to be
+            exceeded
+
+    ***************************************************************************/
+
+    public void opAddAssign ( uint rhs )
+    {
+        if (add(this.payload[0], rhs))
+            enforce(.wideint_exception, checkAndInc(1));
+    }
+
+    unittest
+    {
+        WideInt num = 0;
+        for (auto i = 0; i < 1000; ++i)
+            num += (1 << 31);
+        test!("==")(num.payload[0], 0);
+        test!("==")(num.payload[1], 500);
+    }
+
+    /***************************************************************************
+
+        Mutates current number in-place, dividing it by `rhs` and calculating
+        remainder.
+
+        Params:
+            rhs = number to divide by
+
+        Returns:
+            remainder of division
+
+    ***************************************************************************/
+
+    public uint divideBy ( uint rhs )
+    {
+        ulong remainder = 0;
+
+        for (ptrdiff_t idx = this.payload.length - 1; idx >= 0; --idx)
+        {
+            remainder = (remainder << 32) + this.payload[idx];
+            ulong result = remainder / rhs;
+            remainder -= rhs * result;
+            assert(result <= uint.max);
+            this.payload[idx] = cast(uint) result;
+        }
+
+        assert(remainder <= uint.max);
+        return cast(uint) remainder;
+    }
+
+    unittest
+    {
+        WideInt num = 56 * 47;
+        test!("==")(0, num.divideBy(47));
+        test!("==")(num.payload[0], 56);
+
+        num.payload[0] = 1337;
+        num.payload[1] = 100;
+        test!("==")(1337, num.divideBy(1 << 31));
+        test!("==")(num.payload[0], 200);
+    }
+
+    /***************************************************************************
+
+        Mutates this number in place multiplying it by a regular integer
+
+        Params:
+            rhs = value to multiply by
+
+        Throws:
+            WideIntRangeException if resulting value can't fit in current
+            WideInt value range
+
+    ***************************************************************************/
+
+    public void multiplyBy ( uint rhs )
+    {
+        ulong overflow = 0;
+
+        for (size_t i = 0; i < this.payload.length; ++i)
+        {
+            overflow += cast(ulong)(this.payload[i]) * rhs;
+            this.payload[i] = cast(uint) (overflow & uint.max);
+            overflow >>= 32;
+        }
+
+        enforce(.wideint_exception, overflow == 0);
```

Comment:
I am quite confident we want exception in this case. Native integers have wraparound because it is how hardware works and it is fast. In almost all `WideInt` use cases you assume that it is "big enough" for all your values, essentually using it as `BigInt` with fixed size. If such assumption is violated, you want to know about it.